### PR TITLE
Change default argument of copy method

### PR DESCRIPTION
**Pull request by okuta**

### CODE REVIEW

**Comment by niboshi** [comment]
Do you think we also need to fix Chainer to add `order='C'` in each `copy()` call?
Otherwise layouts of resulting arrays might change.
Or, we could just list this as a compatibility-breaking change in the next release of Chainer.
Or, perhaps we could just ignore the change if we don't have to consider memory layout as an external specification.

**Comment by okuta** [comment]
Using `ascontiguousarray` or `copy(order='C')` is better than current code. We can wait to merge this PR until v3. But, these API incompatibility should be fixed for usability.

**Comment by okuta** [comment]
Rel #158

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -338,7 +338,7 @@ cdef class ndarray:
 
     # TODO(okuta): Implement byteswap
 
-    cpdef ndarray copy(self, order='C'):
+    cpdef ndarray copy(self, order='K'):
```
Comment (takagi) [inline cupy/core/core.pyx]: The default value of `numpy.ndarray.copy`'s `order` option is 'C', while that of `numpy.copy` is 'K'.
https://docs.scipy.org/doc/numpy/reference/generated/numpy.copy.html
https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.copy.html#numpy.ndarray.copy

Comment (takagi) [inline cupy/core/core.pyx]: The default value of `numpy.ndarray.copy`'s order option is 'C', while that of `numpy.copy` is 'K'.
https://docs.scipy.org/doc/numpy/reference/generated/numpy.copy.html
https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.copy.html#numpy.ndarray.copy

So, the changes on core.pyx would not be needed.

File: cupy/creation/from_data.py [hunk 2]
```diff
@@ -80,7 +80,7 @@ def ascontiguousarray(a, dtype=None):
 # TODO(okuta): Implement asmatrix
 
 
-def copy(a, order='C'):
+def copy(a, order='K'):
```
Comment (takagi) [inline cupy/creation/from_data.py]: Would you add a test to check `cupy.copy` works as same as `numpy.copy` on the default value of `order` option?

File: cupy/core/core.pyx (hunk 1 above)
Comment (okuta) [inline cupy/core/core.pyx]: Oh, thanks!

File: cupy/creation/from_data.py (hunk 2 above)
Comment (okuta) [inline cupy/creation/from_data.py]: I added!

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -439,7 +439,7 @@ cdef class ndarray:
         if strides.size() == shape.size():
             newarray = self.view()
         else:
-            newarray = self.copy()
+            newarray = ascontiguousarray(self)
```
Comment (takagi) [inline cupy/core/core.pyx]: Do we leave these `ascontiguousarray` changes? The original `self.copy()`, whose default order is `C`, is enough as well as is natural for the intension.

File: cupy/core/core.pyx [hunk 4]
```diff
@@ -2626,7 +2621,7 @@ cpdef ndarray _take(ndarray a, indices, li=None, ri=None, ndarray out=None):
         if li is not None and ri is not None and li == ri:
             a = rollaxis(a, li)
         if out is None:
-            return a[indices].copy()
+            return ascontiguousarray(a[indices])
```
Comment (takagi) [inline cupy/core/core.pyx]: ditto(leave `ascontiguousarray`?)

File: cupy/core/core.pyx [hunk 5]
```diff
@@ -3392,14 +3387,14 @@ cpdef inline tuple _mat_to_cublas_contiguous(ndarray a, Py_ssize_t trans):
             lda = a._shape[0]
         return a, trans, lda
     if not a._c_contiguous:
-        a = a.copy()
+        a = ascontiguousarray(a)
```
Comment (takagi) [inline cupy/core/core.pyx]: ditto(leave ascontiguousarray?)

File: cupy/core/core.pyx [hunk 6]
```diff
@@ -3392,14 +3387,14 @@ cpdef inline tuple _mat_to_cublas_contiguous(ndarray a, Py_ssize_t trans):
...
     return a, 1 - trans, a._strides[0] // a.itemsize
 
 
 @cython.profile(False)
 cpdef inline tuple _to_cublas_vector(ndarray a, Py_ssize_t rundim):
     if a._strides[rundim] < 0:
-        return a.copy(), 1
+        return ascontiguousarray(a), 1
```
Comment (takagi) [inline cupy/core/core.pyx]: ditto(leave ascontiguousarray?)

**Comment by okuta** [comment]
I updated.

**Comment by takagi** [comment]
LGTM for the PR, while we need to fix Chainer as I post https://github.com/chainer/chainer/issues/3047 before merging.

**Comment by takagi** [comment]
When and how to merge this no-compat change? How about the following process?

1. At **v2.0.0b1**, warn in `cupy.copy` that the default value of `order` argument will be changed at v2.0.0 (requiring another PR)
2. At **v2.0.0**, actually merge this PR and https://github.com/chainer/chainer/pull/3054

We can not follow the Processes to Break Backward Compatibilities as it is because the change of this time is not deprecation of an API.

https://docs-cupy.chainer.org/en/stable/compatibility.html#processes-to-break-backward-compatibilities

**Comment by okuta** [comment]
Please merge in v2.0.0rc1 .

**Pull request merged**